docs/rfcs/038-independent-compute-release.md

```diff
@@ -0,0 +1,298 @@
+3. Trigger e2e tests as usual.
+
+### Release flow
+
+```mermaid
+  sequenceDiagram
+
+  actor oncall as Compute on-call person
+  participant neon as neondatabase/neon
+
+  box private
+    participant cloud as neondatabase/cloud
+    participant exts as neondatabase/build-custom-extensions
+    participant infra as neondatabase/infra
+  end
+
+  box cloud
+    participant preprod as Pre-prod control plane
+    participant prod as Production control plane
+  end
+
+  oncall ->> neon: Open release PR into release-compute
+
+  activate neon
+  neon ->> cloud: CI: trigger e2e compatibility tests
+  activate cloud
+  cloud -->> neon: CI: e2e tests pass
+  deactivate cloud
+  neon ->> neon: CI: pass PR checks, get approvals
+  deactivate neon
+
+  oncall ->> neon: Merge release PR into release-compute
+
+  activate neon
+  neon ->> neon: CI: pass checks, build and push images
+  neon ->> exts: CI: trigger extensions build
+  activate exts
+  exts -->> neon: CI: extensions are ready
+  deactivate exts
+  neon ->> neon: CI: create release tag
+  neon ->> infra: Trigger release workflow using the produced tag
+  deactivate neon
+
+  activate infra
+  infra ->> infra: CI: pass checks
+  infra ->> preprod: Release new compute image to pre-prod automatically <br/> POST /management/api/v2/compute_releases
+  activate preprod
+  preprod -->> infra: 200 OK
+  deactivate preprod
+
+  infra ->> infra: CI: wait for per-region production deploy approvals
+  oncall ->> infra: CI: approve deploys region by region
+  infra ->> prod: POST /management/api/v2/compute_releases
+  activate prod
+  prod -->> infra: 200 OK
+  deactivate prod
+  deactivate infra
+```
+
+## Further work
+
+As briefly mentioned in other sections, eventually we would like to use more complex deployment strategies.
+For example, we can pass a fraction of the total compute starts that should use the new release. Then we can
+mark release as `partial` or `canary` and monitor its performance. If everything is fine, we can promote it
+to `deployed` status. If not, we can rollback to the previous one.
+
+## Alternatives
+
+In theory, we can try using just Helm as is.
+
+1. Write a compute Helm chart. That will actually have only some config map, which control plane can access and read.
+   N.B. We could reuse the control plane chart as well, but then it's not a fully independent release again and even more fuzzy.
+2. Control plane will read it and start using new compute version for new starts.
+
+Drawbacks:
+
+1. Helm releases work best if the workload is controlled by the Helm chart itself. Then you can have different
+   deployment strategies like rolling update or canary or blue/green deployments. At Neon, the compute starts are controlled
+   by control plane, so it makes it much more tricky.
+2. Releases visibility will suffer, i.e. instead of a nice table in control plane and Admin UI, we would need to use
+   `helm` cli and/or K8s UIs like K8sLens.
```

Comment:
On the other hand, we already use helm for the other components, so this would make compute releases more similar to others.

The Admin UI could fetch the information from helm objects too.

Reply:
I put Helm as an alternative because I do think it's doable, but it will be pretty messy, so the price is way too high without clear benefits

One thing is what Muhammet mentioned -- we need access to the previously deployed revisions and their manifests (added a paragraph into RFC)

Another is that normally Helm-charts mount config maps to the service deployed with the same chart, which is not what we want. Yes, we can have some persistent volume or CRD that will keep all the history, but that is tricky again

Some shared configuration service like etcd is also an option, but it's not about Helm already

> Yeah, Helm's advanced deployment strategies won't help us. We'd need to implement that stuff separately in the control plane anyway.

Yeah, we need to have a lot of logic about computes in the control plane anyway (or some other service responsible for provisioning), so I don't see enough reasons for separating computes metadata from it